### PR TITLE
Extract submission eligibility logic to separate library

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Refs: https://github.com/keep-network/keep-core/issues/2725

Submission eligibility logic currently used in relay entry can be reused in DKG. Because of that, we extract that logic to a separate library called `Submission`. That library is deployed as a linked library so we can save on `RandomBeacon` contract size by the way.

### CODE REVIEW

**Comment by nkuba** [comment]
The change looks very reasonable, but the costs worry me as they seem to increase 😕 

### Before this change (main):

```
·--------------------------|-------------·
|  Contract Name           ·  Size (KB)  │
···························|··············
|  RandomBeacon            ·     17.948  │
|  Relay                   ·      0.348  │
|  RandomBeaconStub        ·     18.599  │
```

```
|  Methods                                                                                                                  │
·····················|······························|·············|·············|·············|···············|··············
|  Contract          ·  Method                      ·  Min        ·  Max        ·  Avg        ·  # calls      ·  eur (avg)  │
·····················|······························|·············|·············|·············|···············|··············
|  RandomBeaconStub  ·  submitRelayEntry            ·     245438  ·     262110  ·     253774  ·            2  ·          -  │
·····················|······························|·············|·············|·············|···············|··············
|  Deployments                                      ·                                         ·  % of limit   ·             │
····················································|·············|·············|·············|···············|··············
|  BLS                                              ·          -  ·          -  ·     761370  ·        2.5 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  DKG                                              ·          -  ·          -  ·    1965043  ·        6.6 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  RandomBeaconStub                                 ·          -  ·          -  ·    4585655  ·       15.3 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  RelayStub                                        ·          -  ·          -  ·     528060  ·        1.8 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  SortitionPool                                    ·          -  ·          -  ·    1654539  ·        5.5 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  StakingStub                                      ·          -  ·          -  ·     304263  ·          1 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  TestToken                                        ·          -  ·          -  ·    1271349  ·        4.2 %  ·          -  │
·---------------------------------------------------|-------------|-------------|-------------|---------------|-------------·
```


### After this change:

```
·--------------------------|-------------·
|  Contract Name           ·  Size (KB)  │
···························|··············
|  RandomBeacon            ·     18.159  │
|  Relay                   ·      0.348  │
|  Submission              ·      1.332  │
|  RandomBeaconStub        ·     18.810  │

```

```
|  Methods                                                                                                                  │
·····················|······························|·············|·············|·············|···············|··············
|  Contract          ·  Method                      ·  Min        ·  Max        ·  Avg        ·  # calls      ·  eur (avg)  │
·····················|······························|·············|·············|·············|···············|··············
|  RandomBeaconStub  ·  submitRelayEntry            ·     266952  ·     284087  ·     275520  ·            2  ·          -  │
·····················|······························|·············|·············|·············|···············|··············
|  Deployments                                      ·                                         ·  % of limit   ·             │
····················································|·············|·············|·············|···············|··············
|  BLS                                              ·          -  ·          -  ·     761370  ·        2.5 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  DKG                                              ·          -  ·          -  ·    1965043  ·        6.6 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  RandomBeaconStub                                 ·          -  ·          -  ·    4631113  ·       15.4 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  RelayStub                                        ·          -  ·          -  ·     259404  ·        0.9 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  SortitionPool                                    ·          -  ·          -  ·    1654539  ·        5.5 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  StakingStub                                      ·          -  ·          -  ·     304251  ·          1 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  Submission                                       ·          -  ·          -  ·     340939  ·        1.1 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  TestToken                                        ·          -  ·          -  ·    1271349  ·        4.2 %  ·          -  │
·---------------------------------------------------|-------------|-------------|-------------|---------------|-------------·
```

**Comment by lukasz-zimnoch** [comment]
> 087

Most likely, this is because we use `Submission` as a linked library. We can switch to an internal one but this can increase contract size. Anyway, `submitRelayEntry` will be refactored very soon and all eligibility checks will be removed from here so we can hold with any changes until then.

**Comment by nkuba** [comment]
> Most likely, this is because we use Submission as a linked library. We can switch to an internal one but this can increase contract size.

But contract size has also increased. And this confuses me the most as both transaction cost and contract size are bigger.

**Comment by lukasz-zimnoch** [comment]
Yeah, that's weird. Here is the full comparison of `submitRelayEntry` costs using a higher number of samples:
|                  | Min    | Max    | Average | Calls |
|------------------|--------|--------|---------|-------|
| main             | 245438 | 798657 | 424573  | 35    |
| linked library   | 267001 | 827715 | 447681  | 35    |
| embedded library | 245424 | 798646 | 460371  | 35    |

And the contract size:
|                  | Size   |
|------------------|--------|
| main             | 17.948 |
| linked library   | 18.169 |
| embedded library | 18.058 |

**Comment by lukasz-zimnoch** [comment]
@nkuba Switched `Submission.sol` to be an embedded library.

**Review by nkuba — APPROVED** [review]

**Comment by nkuba** [comment]
I left one non-blocking comment.

### Before this change (main):

```
·--------------------------|-------------·
|  Contract Name           ·  Size (KB)  │
···························|··············
|  RandomBeacon            ·     17.948  │
|  Relay                   ·      0.348  │
|  RandomBeaconStub        ·     18.599  │
```

```
|  Methods                                                                                                                  │
·····················|······························|·············|·············|·············|···············|··············
|  Contract          ·  Method                      ·  Min        ·  Max        ·  Avg        ·  # calls      ·  eur (avg)  │
·····················|······························|·············|·············|·············|···············|··············
|  RandomBeaconStub  ·  submitRelayEntry            ·     245438  ·     262110  ·     253774  ·            2  ·          -  │
·····················|······························|·············|·············|·············|···············|··············
|  Deployments                                      ·                                         ·  % of limit   ·             │
····················································|·············|·············|·············|···············|··············
|  BLS                                              ·          -  ·          -  ·     761370  ·        2.5 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  DKG                                              ·          -  ·          -  ·    1965043  ·        6.6 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  RandomBeaconStub                                 ·          -  ·          -  ·    4585655  ·       15.3 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  RelayStub                                        ·          -  ·          -  ·     528060  ·        1.8 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  SortitionPool                                    ·          -  ·          -  ·    1654539  ·        5.5 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  StakingStub                                      ·          -  ·          -  ·     304263  ·          1 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  TestToken                                        ·          -  ·          -  ·    1271349  ·        4.2 %  ·          -  │
·---------------------------------------------------|-------------|-------------|-------------|---------------|-------------·
```


### After this change:

```
·--------------------------|-------------·
|  Contract Name           ·  Size (KB)  │
···························|··············
|  RandomBeacon            ·     18.058  │
|  Relay                   ·      0.348  │
|  Submission              ·      0.086  │
|  RandomBeaconStub        ·     18.709  │

```

```
|  Methods                                                                                                                  │
·····················|······························|·············|·············|·············|···············|··············
|  Contract          ·  Method                      ·  Min        ·  Max        ·  Avg        ·  # calls      ·  eur (avg)  │
·····················|······························|·············|·············|·············|···············|··············
|  RandomBeaconStub  ·  submitRelayEntry            ·     245424  ·     262096  ·     253760  ·            2  ·          -  │
·····················|······························|·············|·············|·············|···············|··············
|  Deployments                                      ·                                         ·  % of limit   ·             │
····················································|·············|·············|·············|···············|··············
|  BLS                                              ·          -  ·          -  ·     761370  ·        2.5 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  DKG                                              ·          -  ·          -  ·    1965043  ·        6.6 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  RandomBeaconStub                                 ·          -  ·          -  ·    4609500  ·       15.4 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  RelayStub                                        ·          -  ·          -  ·     259404  ·        0.9 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  SortitionPool                                    ·          -  ·          -  ·    1654539  ·        5.5 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  StakingStub                                      ·          -  ·          -  ·     304263  ·          1 %  ·          -  │
····················································|·············|·············|·············|···············|··············
|  TestToken                                        ·          -  ·          -  ·    1271349  ·        4.2 %  ·          -  │
·---------------------------------------------------|-------------|-------------|-------------|---------------|-------------·
```

File: solidity/random-beacon/test/Submission.test.ts [hunk 1]
```diff
@@ -0,0 +1,182 @@
+/* eslint-disable @typescript-eslint/no-unused-expressions, no-await-in-loop */
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/test/Submission.test.ts:1]: Adding `await` before `expect` which doesn't deal with a promise can be misleading. We usually don't do that in tests.  However, I agree we may not want to disable the linter for the entire file and disable specific lines instead.

**Pull request merged**